### DISCUSSION TITLE
Adding Customer Code blocks in the Pre- and Post- functions

### DISCUSSION THREAD

**Discussion by smiller0924** [thunderclient/thunder-client-support#955] [Ideas]
I am just becoming familiar with your product.  The majority of my experience is with Postman.

One of the features that I enjoy with Postman is to run custom programming either prior to or after a API call.  

Here are examples of Pre-Request Scripts.  The requirement is that I need to get a date to feed into a variable in a specific format:

`var moment = require('moment');
pm.environment.set('PickupDate', moment().format("YYYY-MM-DD")+"T16:00:00Z");`

This will get the date and then return the value in the format: **_2022-12-30T16:00:00Z_** (Thes function is used to set a pickup request that must be in a specific format required by the carrier.

Another example would be:

`var moment = require('moment')
pm.environment.set("future_timestamp", moment().add(1, 'day').format("YYYY-MM-DD")+"T11:00:00Z")`

This request is similar to the first.  But it advanced the date by one day and returns the request as **_2022-12-31T16:00:00Z_**

Similarly, I need to run Post scripts.  In this example I am decoding a base64 document reply that returns multiple labels in an array:

```js
var allLabelBuffers=[];
var totPackageCount = (jsonData.output.transactionShipments[0].pieceResponses.length)
var packageCount = 0

for (var packageCount; packageCount < totPackageCount; packageCount++) 
                  {
                   var newLabelBuffer = (jsonData.output.transactionShipments[0].pieceResponses[packageCount].packageDocuments[0].encodedLabel)
                     allLabelBuffers.push(atob(newLabelBuffer))
                   }

       allLabelBuffers = JSON.stringify(allLabelBuffers)

pm.environment.set("labelBuffer", allLabelBuffers); 
```

The atob(newLabelBuffer) decodes the base64 to actual document language.  In this case it is a Zebra ZPL printer code.  

Having the ability to run code prior to an API Request and after an API Reply would be extremely helpful.  It would also give me an alternative to Postman should I need it.

Thank you.

**Comment by rangav** [comment]
Hi @smiller0924 
Pre and Post request scripting is possible with filters

see docs here
https://github.com/rangav/thunder-client-support/blob/master/docs/filters.md

let me know if you have any questions

**Reply by smiller0924** [reply]
I have seen this. But how do you execute the script to "push" json.response header values to an environmental variable. If you could provide a better sample as a frame of reference, that would be most helpful. 

Thank you. Happy New Year.

**Reply by rangav** [reply]
Thanks, Happy new year to you.

you can also you filters in Tests, see 3rd test in the image below.
https://github.com/rangav/thunder-client-support/blob/master/docs/filters.md#filters-in-tests

**Reply by rangav** [reply]
Set env variable in tests docs
https://github.com/rangav/thunder-client-support#setenv

**Reply by smiller0924** [reply]
Ok. I think I kind of understand the "Pre-"  

How do I extract specific header values from an array. Ex: I'm getting 5 separate Base64 label replies. 

How do I create a "count" set the "Do While...". Then feed each decoded (atob) ZPL reply separated into an Environmental Variable?

Thank you.

**Reply by rangav** [reply]
I don’t have access to laptop now, will provide example later tomorrow 

but let me try to explain 
You use the custom logic in filter and decode the value and `return` the value in the function 

in tests
Create test using `set env variable` and use the above filter

example `json | customerFilter`

**Reply by smiller0924** [reply]
So I call the custom filter. Then it parses the Jason response. Then my script gets executed. The values in for Do While Loop populate my environmental variable. 

If that the basic idea?  If so, an example would be helpful toward pointing me in the right direction. 

Thank you very much.

**Comment by rangav** [comment]
Here are the steps

1. Create script file
<img width="1103" alt="Screenshot 2023-01-02 at 10 55 23" src="https://user-images.githubusercontent.com/8637550/210223805-a1c235dc-851e-4fa1-b8cf-69675dc2f44c.png">

2. Attach script to collection settings
<img width="1241" alt="Screenshot 2023-01-02 at 10 56 04" src="https://user-images.githubusercontent.com/8637550/210223851-75d204ad-e9e0-4f1b-bf31-e8fbca414574.png">

3. Create a test using `Set Env Variable` option
<img width="780" alt="Screenshot 2023-01-02 at 10 57 31" src="https://user-images.githubusercontent.com/8637550/210223904-9d1d5daa-a35f-46d3-878c-c27c98b3c411.png">

4. Apply the custom filter `decodeFilter`
 - The response body is passed to filter using the json keyword
<img width="788" alt="Screenshot 2023-01-02 at 10 58 27" src="https://user-images.githubusercontent.com/8637550/210223966-28fe2408-2d46-42bf-ab04-e8c066e2c468.png">

6. Create an environment, if you haven't already created.
<img width="271" alt="Screenshot 2023-01-02 at 11 12 05" src="https://user-images.githubusercontent.com/8637550/210224032-5c059593-fd4a-44db-baef-01e072741db2.png">

6 Run the request and see if its working or not, see also `logs` for any errors

**Reply by smiller0924** [reply]
Thank you for this. This is very helpful. 

I do have one question. I noticed you crossed out the "atob" that decodes my Base64 reply. Why did that need to be done?  Shouldn't I run the decode in the script and send the unencoded header value to the environmental variable?  

Please clarify.  Thank you again.

**Reply by rangav** [reply]
Vscode did that not me, just saying it’s deprecated in node.

**Reply by smiller0924** [reply]
Ok. One more question. Which library are you running to generate the node.js. The examples I provided are from Postman. Which I believe uses 'C'.  

Sorry for all the questions. Just trying to get a good handle on this function. 

Thank you.

**Reply by smiller0924** [reply]
Sorry. The code is in Java. Not 'C'.

**Reply by rangav** [reply]
The extension is built on node.js and JavaScript.

the same code should work in Thunder Client too.

Have you tested the script?

**Comment by smiller0924** [comment]
Ok.  I have tested you suggestions.  It is not parsing out the Label Buffers.  It is instead returning the entire JSON Response Body as well as the filters.js as you recommended it be written.  (I have zipped the response as the upload does not support .json or .js extensions).  It does not appear to be running the script at all from the function from the filters.js.  In fact it is ignoring the filter is even there.  

Any thoughts you have would be helpful.  Thank you.


![image](https://user-images.githubusercontent.com/108164566/210420763-ac4dec70-88ef-451a-957d-6356beba318e.png)

[filters.zip](https://github.com/rangav/thunder-client-support/files/10338987/filters.zip)
[REST_RESPONSE_BODY.zip](https://github.com/rangav/thunder-client-support/files/10338956/REST_RESPONSE_BODY.zip)

**Comment by smiller0924** [comment]
I did add the json. in front of "output...".  Still no result:

![image](https://user-images.githubusercontent.com/108164566/210422827-d775f543-f2f7-4672-b92f-0e0c054f8ac7.png)

**Reply by rangav** [reply]
Thanks for sharing response body, will test and get back to you with a solution

**Reply by rangav** [reply]
How did you unzip the response in Postman? Can you share complete script you used?

**Reply by smiller0924** [reply]
Here is the exact script from Postman Test module:

**`_var jsonData = JSON.parse(responseBody);_**

pm.environment.set("shipDate", jsonData.output.transactionShipments[0].shipDatestamp);
pm.environment.set("serviceType", jsonData.output.transactionShipments[0].serviceType);

var allTrackingNumbers=[];
var totPackageCount = (jsonData.output.transactionShipments[0].pieceResponses.length)
var packageCount = 0

for (var packageCount; packageCount < totPackageCount; packageCount++) 
                  {
                   var trackingNo = (jsonData.output.transactionShipments[0].pieceResponses[packageCount].trackingNumber)
                     allTrackingNumbers.push(trackingNo)
                   }

       allTrackingNumbers = JSON.stringify(allTrackingNumbers)

pm.environment.set("trackingNo", allTrackingNumbers); 

**_var allLabelBuffers=[];
var totPackageCount = (jsonData.output.transactionShipments[0].pieceResponses.length)
var packageCount = 0

for (var packageCount; packageCount < totPackageCount; packageCount++) 
                  {
                   var newLabelBuffer = (jsonData.output.transactionShipments[0].pieceResponses[packageCount].packageDocuments[0].encodedLabel)
                     allLabelBuffers.push(atob(newLabelBuffer))
                   }

       allLabelBuffers = JSON.stringify(allLabelBuffers)

pm.environment.set("labelBuffer", allLabelBuffers);_** `

Needless to say the Bold/Italicized sections are what I have been focusing on with you.  Here is an image from Postman:

![image](https://user-images.githubusercontent.com/108164566/210433320-65fba8a9-80d1-48ba-8d3d-5163798d6ae0.png)


I hope this helps.  Let me know if you have any questions.

**Comment by smiller0924** [comment]
One other thing I noticed.  I get a 'lightbulb' by the line, "exports.decodeFilter = decodeFilter;"  It make a recommendation to, "Convert to an ES Module".  If I select that then it conversts "exports.decodeFilter = decodeFilter;" to:

const _decodeFilter = decodeFilter;
export { _decodeFilter as decodeFilter };

I've tried that modification.  It seems almost as if the "json | decodeFilter" is completely ignoring the "decodeFilter" and just pushing the entire reply into my variable.

**Reply by rangav** [reply]
you should only use `module.exports = [decodeFilter]`, our internal code parser expects this.

**Reply by smiller0924** [reply]
Ok.  This is exactly what I have in my filters.js:

`function decodeFilter(json) {
    var allLabelBuffers = [];
    var totPackageCount = (json.output.transactionShipments[0].pieceResponses.length);
    var packageCount = 0;

    for (var packageCount; packageCount < totPackageCount; packageCount++) {
        var newLabelBuffer = (json.output.transactionShipments[0].pieceResponses[packageCount].packageDocuments[0].encodedLabel);
        allLabelBuffers.push(atob(newLabelBuffer));
    }

    allLabelBuffers = JSON.stringify(allLabelBuffers);

    //pm.environment.set("labelBuffer", allLabelBuffers); 
    return allLabelBuffers;

}
module.exports = [decodeFilter];`

**Comment by rangav** [comment]
Its working now, `atob` module is not available in Thunder Client.
<img width="978" alt="Screenshot 2023-01-03 at 21 05 31" src="https://user-images.githubusercontent.com/8637550/210441468-9cd272ad-3eed-4fe3-99f9-f0cc52bf06e8.png">


Code:

```js
function decodeFilter(json) {
    var allLabelBuffers = [];
    var totPackageCount = (json.output.transactionShipments[0].pieceResponses.length)
    var packageCount = 0

    for (var packageCount; packageCount < totPackageCount; packageCount++) {
        var newLabelBuffer = (json.output.transactionShipments[0].pieceResponses[packageCount].packageDocuments[0].encodedLabel)
        allLabelBuffers.push(Buffer.from(newLabelBuffer).toString())
    }

    console.log("allLabelBuffers", allLabelBuffers);

    allLabelBuffers = JSON.stringify(allLabelBuffers)

    // pm.environment.set("labelBuffer", allLabelBuffers);
    return allLabelBuffers;
}

module.exports = [decodeFilter];
```

**Reply by smiller0924** [reply]
Ok.  That does work.  Now how do I decode the Base64?  I saw this reference in your documentation.  However, I am uncertain how to execute this:

atob
The filter will decode the value, The syntax is {{input | atob}}
if input value is SGVsbG8gV29ybGQ=, then result is Hello World

Could you elaborate please?

**Reply by rangav** [reply]
since you already using custom script, you can decode in the script itself.

`allLabelBuffers.push(Buffer.from(newLabelBuffer, "base64").toString())`

**Reply by rangav** [reply]
here is the updated script with `base64` decode:

```js
function decodeFilter(json) {
    var allLabelBuffers = [];
    var totPackageCount = (json.output.transactionShipments[0].pieceResponses.length)
    var packageCount = 0

    for (var packageCount; packageCount < totPackageCount; packageCount++) {
        var newLabelBuffer = (json.output.transactionShipments[0].pieceResponses[packageCount].packageDocuments[0].encodedLabel)
        allLabelBuffers.push(Buffer.from(newLabelBuffer, "base64").toString())
    }

    allLabelBuffers = JSON.stringify(allLabelBuffers)

    // pm.environment.set("labelBuffer", allLabelBuffers);
    return allLabelBuffers;
}

module.exports = [decodeFilter];

```

**Reply by smiller0924** [reply]
That got it!  Thank you so much for working with me.  I understand you product much better.  This is seriously cool!

Thank you again and have a Happy New Year!

**Reply by rangav** [reply]
You are welcome, Happy New Year 🎉 

Feel free to reach out if you have any further questions.

**Reply by smiller0924** [reply]
Like I said this was definitely cool.  I'm looking for an alternative to Postman.  This definitely got my vote!

**Comment by smiller0924** [comment]
I had another thought (yes, that's a dangerous thing with me).  Can I use one of the Post-Script events to create a .js script that would print my decoded ZPL buffer directly to a printer instead of sending it to a variable?

**Reply by rangav** [reply]
Did not understood - `directly to a printer`? Can you please provide more details

**Reply by smiller0924** [reply]
Sorry.  I'll try to elaborate.  I would like to take the decoded Base64 ZPL String and send it to a printer that is connected to my PC.  So instead of feeding the decoded ZPL Strings to a variable I'd like to run a script with a printer path and print command from the Response advance script.  If that possible?

**Reply by smiller0924** [reply]
Here is a website that outlines such an action: https://www.neodynamic.com/articles/How-to-print-raw-Zebra-ZPL-commands-from-Javascript/

**Reply by smiller0924** [reply]
Here is a better example: https://stackoverflow.com/questions/36570689/javascript-send-raw-text-to-printer-no-server-requests-method-calls-able-to

`<script type="text/javascript">
  function openWin() {
    var printWindow = window.open();
    printWindow.document.open('text/plain')
    printWindow.document.write('${^XA^FO50,100^BXN,10,200^FDYourTextHere^FS^XZ}$');
    printWindow.document.close();
    printWindow.focus();
    printWindow.print();
  }
</script>
<input type="button" value="Print code" onclick="openWin()" />`

**Reply by rangav** [reply]
Thanks for sharing details? did you do this in Postman?

**Reply by smiller0924** [reply]
Postman cannot execute printing outside of the application.  The best I can do is call an API from Labelary to display a label image.  I'd like to push the label to a shared printer connected to my PC and Postman does not support this.

**Reply by rangav** [reply]
you can use if there is any `node` modules that support this.

To import node module
https://github.com/rangav/thunder-client-support/blob/master/docs/filters.md#import-node-module

**Reply by smiller0924** [reply]
What would that look like in the Test section.  How does it get triggered?

**Reply by rangav** [reply]
You need to create a `dummy/fake` test for this purpose

same steps as here
https://github.com/rangav/thunder-client-support/discussions/955#discussioncomment-4570843

Instead of selecting `Set Env Variable`, you can select `Json Query` 

Your custom script just returns the same `json` back

<img width="744" alt="Screenshot 2023-01-04 at 18 01 27" src="https://user-images.githubusercontent.com/8637550/210620205-9a2451db-e871-410f-a3e8-8f6d52d46999.png">

**Reply by smiller0924** [reply]
Ok.  I'll play with that.  Thank you.
